### PR TITLE
Refactor code for creating Cacher.

### PR DESCRIPTION
**Pull request by wojtek-t**
Requested in #16496

### CODE REVIEW

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: pkg/registry/pod/etcd/etcd.go [hunk 1]
```diff
@@ -64,23 +64,14 @@ func NewStorage(s storage.Interface, useCacher bool, k client.ConnectionInfoGett
...
-			Storage:        s,
-			Type:           &api.Pod{},
-			ResourcePrefix: prefix,
-			KeyFunc: func(obj runtime.Object) (string, error) {
-				return storage.NamespaceKeyFunc(prefix, obj)
-			},
-			NewListFunc: func() runtime.Object { return &api.PodList{} },
-		}
-		storageInterface = storage.NewCacher(config)
+		storageInterface = storage.NewCacher(1000, s, nil, &api.Pod{}, prefix, true, newListFunc)
```
Comment (lavalamp) [inline pkg/registry/pod/etcd/etcd.go]: I was actually hoping to construct the cacher outside and pass it in-- if that's too hard, then I'll take this, though.

Comment (wojtek-t) [inline pkg/registry/pod/etcd/etcd.go]: It's different for every resource, so yeah - that would be difficult.

Comment (lavalamp) [inline pkg/registry/pod/etcd/etcd.go]: (Discussed in person: I suggested passing a construction function in instead of the boolean)

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

[Any better suggestions for names are appreciated]

File: pkg/registry/pod/etcd/etcd.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/registry/pod/etcd/etcd.go]: Done.

File: pkg/storage/cacher.go [hunk 2]
```diff
@@ -111,10 +111,56 @@ type Cacher struct {
 	ListFromCache bool
 }
 
+type DecorateStorage func(
+	Interface, int, Versioner, runtime.Object, string, bool,
+	func() runtime.Object) Interface
```
Comment (lavalamp) [inline pkg/storage/cacher.go]: Can you name the parameters? + add godoc.

I might call this a storage factory...

Comment (wojtek-t) [inline pkg/storage/cacher.go]: Done.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

**Comment by lavalamp** [comment]
LGTM, thanks for bearing with me :)

**Pull request merged**